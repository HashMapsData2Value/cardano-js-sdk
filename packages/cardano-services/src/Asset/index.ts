export * from './AssetBuilder';
export * from './AssetHttpService';
export * from './CardanoTokenRegistry';
export * from './DbSyncAssetProvider';
export * from './DbSyncNftMetadataService';
export * from './types';
