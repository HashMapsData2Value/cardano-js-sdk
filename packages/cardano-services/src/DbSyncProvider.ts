import { CardanoNode, HealthCheckResponse, Provider } from '@cardano-sdk/core';
import { Pool } from 'pg';

const HEALTH_CHECK_QUERY = 'SELECT 1';
// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyArgs = any[];
export const DbSyncProvider = <
  T extends (abstract new (...args: AnyArgs) => {}) | (new (...args: AnyArgs) => {}) = { new (): {} }
>(
  BaseClass?: T
) => {
  abstract class Mixin extends (BaseClass || Object) implements Provider {
    public db: Pool;
    public cardanoNode: CardanoNode;

    constructor(...args: AnyArgs) {
      const [db, cardanoNode, ...baseArgs] = [...args];

      super(...baseArgs);

      this.db = db;
      this.cardanoNode = cardanoNode;
    }

    public async healthCheck(): Promise<HealthCheckResponse> {
      // TODO: query block table, get last block minus 5
      const dbHealthCheck = { ok: !!(await this.db.query(HEALTH_CHECK_QUERY)).rowCount };
      const cardanoNodeHealthCheck = await this.cardanoNode.healthCheck();
      return {
        localNode: cardanoNodeHealthCheck.localNode ?? undefined,
        ok: dbHealthCheck.ok && cardanoNodeHealthCheck.ok
      };
    }
  }

  type BaseArgs = T extends abstract new (...baseArgs: infer A) => {}
    ? A
    : T extends new (...baseArgs: infer A) => {}
    ? A
    : never;
  type BaseInstance = T extends abstract new (...baseArgs: AnyArgs) => infer I
    ? I
    : T extends new (...baseArgs: AnyArgs) => infer I
    ? I
    : never;
  type ReturnedType = BaseInstance & {
    db: Pool;
    cardanoNode: CardanoNode;
    healthCheck: () => Promise<HealthCheckResponse>;
  };

  return Mixin as unknown as (T extends new (...baseArgs: AnyArgs) => {}
    ? new (db: Pool, cardanoNode: CardanoNode, ...args: BaseArgs) => ReturnedType
    : abstract new (db: Pool, cardanoNode: CardanoNode, ...args: BaseArgs) => ReturnedType) & {
    prototype: { healthCheck: () => Promise<HealthCheckResponse> };
  };
};
