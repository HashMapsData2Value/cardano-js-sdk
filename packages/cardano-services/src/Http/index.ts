export * from './HttpServer';
export * from './HttpService';
