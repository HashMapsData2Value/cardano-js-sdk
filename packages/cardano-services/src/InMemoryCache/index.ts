export * from './InMemoryCache';
export * from './defaults';
