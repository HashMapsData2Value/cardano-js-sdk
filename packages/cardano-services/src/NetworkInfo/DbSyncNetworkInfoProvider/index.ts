export * from './DbSyncNetworkInfoProvider';
export * from './utils';
export * as NetworkInfoCacheKey from './keys';
