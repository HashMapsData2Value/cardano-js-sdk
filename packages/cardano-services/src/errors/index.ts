export * from './InvalidLoggerLevel';
export * from './InvalidModuleState';
