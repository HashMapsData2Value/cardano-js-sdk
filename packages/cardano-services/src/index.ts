export * from './Http';
export * from './Program';
export * from './RunnableModule';
export * from './StakePoolSearch';
export * from './TxSubmit';
export * from './TxWorker';
export * from './Utxo';
export * as errors from './errors';
