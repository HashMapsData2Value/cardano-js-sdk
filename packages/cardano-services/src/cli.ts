#!/usr/bin/env node
/* eslint-disable import/imports-first */
require('../scripts/patchRequire');
import {
  API_URL_DEFAULT,
  OGMIOS_URL_DEFAULT,
  ProgramOptionDescriptions,
  ServiceNames,
  loadHttpServer
} from './Program';
import { Command } from 'commander';
import { InvalidLoggerLevel } from './errors';
import { LogLevel } from 'bunyan';
import { URL } from 'url';
import { loggerMethodNames } from './util';
import onDeath from 'death';
const clear = require('clear');
const packageJson = require('../package.json');

clear();
// eslint-disable-next-line no-console
console.log('Cardano Services CLI');

const program = new Command('cardano-services');

program.version(packageJson.version);

program
  .command('start-server')
  .description('Start the HTTP server')
  .argument('<serviceNames...>', `List of services to attach: ${Object.values(ServiceNames).toString()}`)
  .option('--api-url <apiUrl>', ProgramOptionDescriptions.ApiUrl, (url) => new URL(url), new URL(API_URL_DEFAULT))
  .option('--enable-metrics <metricsEnabled>', ProgramOptionDescriptions.MetricsEnabled, false)
  .option('--db-connection-string <dbConnectionString>', ProgramOptionDescriptions.DbConnection, (url) =>
    new URL(url).toString()
  )
  .option(
    '--ogmios-url <ogmiosUrl>',
    ProgramOptionDescriptions.OgmiosUrl,
    (url) => new URL(url),
    new URL(OGMIOS_URL_DEFAULT)
  )
  .option(
    '--logger-min-severity <level>',
    ProgramOptionDescriptions.LoggerMinSeverity,
    (level) => {
      if (!loggerMethodNames.includes(level)) {
        throw new InvalidLoggerLevel(level);
      }
      return level;
    },
    'info'
  )
  .action(
    async (
      serviceNames: ServiceNames[],
      {
        apiUrl,
        dbConnectionString,
        loggerMinSeverity,
        metricsEnabled,
        ogmiosUrl
      }: {
        apiUrl: URL;
        dbConnectionString?: string;
        loggerMinSeverity: LogLevel;
        metricsEnabled: boolean;
        ogmiosUrl?: URL;
      }
    ) => {
      const server = await loadHttpServer({
        apiUrl: apiUrl || API_URL_DEFAULT,
        options: {
          dbConnectionString,
          loggerMinSeverity,
          metricsEnabled,
          ogmiosUrl
        },
        serviceNames
      });
      await server.initialize();
      await server.start();
      onDeath(async () => {
        await server.shutdown();
        process.exit(1);
      });
    }
  );

if (process.argv.slice(2).length === 0) {
  program.outputHelp();
  process.exit(1);
} else {
  program.parseAsync(process.argv).catch((error) => {
    // eslint-disable-next-line no-console
    console.error(error);
    process.exit(0);
  });
}
