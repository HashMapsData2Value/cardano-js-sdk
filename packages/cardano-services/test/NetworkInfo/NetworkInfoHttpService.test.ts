/* eslint-disable @typescript-eslint/no-explicit-any */
import { Cardano, NetworkInfo, NetworkInfoProvider, testnetTimeSettings } from '@cardano-sdk/core';
import { DbSyncNetworkInfoProvider, NetworkInfoHttpService } from '../../src/NetworkInfo';
import { HttpServer, HttpServerConfig } from '../../src';
import { Pool } from 'pg';
import { getPort } from 'get-port-please';
import { networkInfoHttpProvider } from '@cardano-sdk/cardano-services-client';
import axios from 'axios';

const UNSUPPORTED_MEDIA_STRING = 'Request failed with status code 415';
const APPLICATION_CBOR = 'application/cbor';
const APPLICATION_JSON = 'application/json';

describe('NetworkInfoHttpService', () => {
  let dbConnection: Pool;
  let httpServer: HttpServer;
  let networkInfoProvider: DbSyncNetworkInfoProvider;
  let service: NetworkInfoHttpService;
  let port: number;
  let apiUrlBase: string;
  let config: HttpServerConfig;
  let cardanoNodeConfigPath: string;

  beforeAll(async () => {
    port = await getPort();
    config = { listen: { port } };
    apiUrlBase = `http://localhost:${port}/network-info`;
    cardanoNodeConfigPath = process.env.CARDANO_NODE_CONFIG_PATH!;
    dbConnection = new Pool({ connectionString: process.env.DB_CONNECTION_STRING });
  });

  afterEach(async () => {
    jest.resetAllMocks();
  });

  describe('healthy state', () => {
    beforeAll(async () => {
      networkInfoProvider = new DbSyncNetworkInfoProvider(cardanoNodeConfigPath, dbConnection);
      service = NetworkInfoHttpService.create({ networkInfoProvider });
      httpServer = new HttpServer(config, { services: [service] });
      await httpServer.initialize();
      await httpServer.start();
    });

    afterAll(async () => {
      await dbConnection.end();
      await httpServer.shutdown();
    });

    describe('/health', () => {
      it('forwards the networkInfoProvider health response', async () => {
        const res = await axios.post(`${apiUrlBase}/health`, {
          headers: { 'Content-Type': APPLICATION_JSON }
        });
        expect(res.status).toBe(200);
        expect(res.data).toEqual({ ok: true });
      });
    });

    describe('/network', () => {
      it('returns a 200 coded response with a well formed HTTP request', async () => {
        expect((await axios.post(`${apiUrlBase}/network`, { args: [] })).status).toEqual(200);
      });

      it('returns a 415 coded response if the wrong content type header is used', async () => {
        try {
          await axios.post(`${apiUrlBase}/network`, { args: [] }, { headers: { 'Content-Type': APPLICATION_CBOR } });
        } catch (error: any) {
          expect(error.response.status).toBe(415);
          expect(error.message).toBe(UNSUPPORTED_MEDIA_STRING);
        }
      });

      describe('with NetworkInfoHttpProvider', () => {
        let provider: NetworkInfoProvider;
        beforeEach(() => {
          provider = networkInfoHttpProvider(apiUrlBase);
        });

        it('time settings response is an array of network info response', async () => {
          const testnetNetworkInfo: NetworkInfo['network'] = {
            id: Cardano.NetworkId.testnet,
            magic: Cardano.CardanoNetworkMagic.Testnet,
            timeSettings: testnetTimeSettings
          };

          const response = await provider.networkInfo();
          expect(response.network).toEqual(testnetNetworkInfo);
        });

        it('response is an object of network info', async () => {
          const response = await provider.networkInfo();
          expect(response).toMatchSnapshot();
        });
      });
    });
  });
});
