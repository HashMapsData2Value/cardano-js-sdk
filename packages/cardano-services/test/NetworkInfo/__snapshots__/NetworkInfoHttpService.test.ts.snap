// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkInfoHttpService healthy state /network with NetworkInfoHttpProvider response is an object of network info 1`] = `
Object {
  "lovelaceSupply": Object {
    "circulating": 69163989759711432n,
    "max": 45000000000000000n,
    "total": 40342337171803411n,
  },
  "network": Object {
    "id": 0,
    "magic": 1097911063,
    "timeSettings": Array [
      Object {
        "epochLength": 21600,
        "fromSlotDate": 2019-07-24T20:20:16.000Z,
        "fromSlotNo": 0,
        "slotLength": 20000,
      },
      Object {
        "epochLength": 432000,
        "fromSlotDate": 2020-07-28T19:20:16.000Z,
        "fromSlotNo": 1598400,
        "slotLength": 1000,
      },
    ],
  },
  "stake": Object {
    "active": 14977960053990832n,
    "live": 69154425288708824n,
  },
}
`;
